core/src/main/java/org/elasticsearch/common/lucene/search/Queries.java

```diff
@@ -179,8 +179,6 @@ public static int calculateMinShouldMatch(int optionalClauseCount, String spec)
             result = calc < 0 ? result + calc : calc;
         }
 
-        return (optionalClauseCount < result ?
-                optionalClauseCount : (result < 0 ? 0 : result));
-
+        return result < 0 ? 0 : result;
```

Comment:
The boolean query throws an IllegalArgumentException when min should match is a negative value but only on some cases, for instance if you don't have should clauses then the query works.
I can add more tests in order to emphasize the kind of problem we could have with this approach but I think we should first define what is a valid min should match value first. For instance what should we do if min should match is set but there is no "should" clause ?